keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1182 @@
+
+Benefits to upgrading binary version independent of compatibility version:
+
+- A skip-version binary upgrade that transitions through compat versions has
+  higher probability of bugs in those intermediate binary versions already being
+  fixed.
+- During an upgrade, it is possible upgrade the new binary version while the
+  compatibility version is fixed (e.g. `binaryVersion: 1.28` -> `{binaryVersion: 1.29, compatVersion: 1.28}`).
+  This allow differences in the binary (bugs fixed or introduced, performance
+  changes, ...) to be introduced into a cluster and verified safe before
+  allowing access to the new APIs and features new version. Once the binary
+  version has been deamed safe, the compatibility version can then be upgraded.
+- Any upgrade system that successfully detects failures between upgrade steps
+  can report which upgrade step failed. This makes it easier to diagnose the
+  failures, because there are fewer possible causes of the failure. (There's a
+  huge difference between "A cluster upgrade failed" and "A cluster upgrade
+  failed when only the apiserver's binary version changed").
+- For upgrades where multiple failures occur, this increases the odds those
+  failures are split across steps. An upgrade system that is able to pause after
+  a step where failures are detected allow for failures at that step to be
+  addressed before proceeding to subsequent steps. These failures can be
+  addressed without the disruption and "noise" from failures in subsequent
+  steps.
+- A compatibility version rollback can be performed without changing binary
+  version.
+
+
+### Goals
+
+- Introduce the metadata necessary to configure features/APIs/storage-versions/CEL
+  features to match the behavior of an older Kubernetes release version
+- A Kubernetes binary with compatibility version set to N, will pass the
+  conformance and e2e tests from Kubernetes release version N.
+- A Kubernetes binary with compatibility version set to N does not enable any
+  changes (storage versions, CEL feature, features) that would prevent it
+  from being rolled back to N-1.
+- The most recent Kubernetes version supports compatibility version being set to
+  the full range of supported versions (N..N-3).
+
+### Non-Goals
+
+- Support compatibility version for Alpha features.  Alpha feature are not
+  designed to be upgradable, so we will not allow alpha features to be used in
+  conjunction with compatibility version.
+- Changes to Cluster API/kubeadm/KIND/minikube to absorb the compatibility versions
+  will be addressed separate from this KEP
+
+## Proposal
+
+### Component Flags
+
+Kubernetes components (apiservers, controller managers, schedulers) will offer a
+`--compatibility-version` flag that can be set to any of the previous three
+minor versions. If unset, the compatibility version defaults to the `<major.minor>`
+version of the binary version.
```

Comment:
> If unset, the compatibility version defaults to the `<major.minor>` version of the binary version.

The default should match the current behavior, but I'm not really sure how I'd describe the current behavior for `kube-apiserver`. There's an internal and external aspect to compatibility mode.

For the "internal" aspect, `kube-apiserver` already maintains n-1 compatibility for persisted data and validation behaviors by default:
* we only persisted versions in etcd the previous server can read
* we only allow data into new API fields once they existed in the previous release
* we only relax validation after the previous release tolerates it

That seems like `kube-apiserver` is operating in 1.(n-1) compatibility mode in some ways. We have to keep doing that by default.

For the "external" aspect, `kube-apiserver` acts like 1.n by default (serves new versions, etc). I'd guess we want to keep doing that by default as well?

How do we reconcile both of these into a single default, and for `kube-apiserver`, should the default be `<major>.<minor>` or `<major>.<minor-1>`? If I set `--compatibility-version=1.30`, am I telling a 1.30 `kube-apiserver` it is ok to persist data a 1.29 `kube-apiserver` couldn't handle?

edit: I just got down to the `StorageVersion Compatibility Versioning` section, and I *think* that is saying "compatibility version 1.n will be compatible from 1.(n-1) to 1.(n+1)". That's a little hard to wrap my head around.

Reply:
because the policy allows for server `--compatibility-version=1.30` to be mixed with servers with  `--compatibility-version=1.29`, so if there is possibility of incompatibility, the logic to save/convert 1.29 data would needed to be preserved for 3 releases, similar to the section of *Changes to Feature Gates*

Reply:
I've struggled with the terms we picked. 

When we picked the term compatibility version, we COULD have either meant:

1. The version a component is backward compatible with (implicitly n-1 today)
2. The version a component emulates (not possible today)

[CEL's internal compatibility version](https://github.com/kubernetes/kubernetes/blob/f07b47c3d1e596f53640285afb0d53a81fe6f259/staging/src/k8s.io/apiserver/pkg/cel/environment/base.go#L45) is an example of (1).

Oracle's [COMPATIBLE feature](https://docs.oracle.com/en/database/oracle/oracle-database/12.2/upgrd/what-is-oracle-database-compatibility.html) is an example of (2).

What we've ended up describing in this KEP is (2).   When I think about versions I mentally use the term "effective version" to think about how an apiserver behaves.  If no compatibility version is set, the effective version IS the binary version.  If a compatibility version is set, the effective version is that compatibility version.

This does have limitations.  We can't, for example, run a component in a mode that doesn't require any backward compatibility with prior versions, which could be handy if running a entirely new cluster that will never need to be rolledback.

Reply:
> This does have limitations. We can't, for example, run a component in a mode that doesn't require any backward compatibility with prior versions, which could be handy if running a entirely new cluster that will never need to be rolledback.

Yeah, at one point we were thinking about the final rollback-safe state of an upgrade being `--compatibility-mode=<binaryversion>` and being able to drop the n-1 storage and validation compatibility requirements which would accelerate by 4 months how quickly things like that could become effective. I'm sad if we can't get the "go faster" benefits of explicitly telling the apiserver it doesn't need to keep n-1 compatibility once an upgrade is rollback-safe.

Reply:
> - we only allow data into new API fields once they existed in the previous release
> - we only relax validation after the previous release tolerates it

I've added a small section about ratcheting since we hadn't discussed it in the KEP.  I've also added both of these to our list of integration tests.

Reply:
> Yeah, at one point we were thinking about the final rollback-safe state of an upgrade being `--compatibility-mode=<binaryversion>` and being able to drop the n-1 storage and validation and cel compatibility requirements which would accelerate by 4 months how quickly improved validation and new cel capabilities could become effective. I'm sad if we can't get the "go faster" benefits of explicitly telling the apiserver it doesn't need to keep n-1 compatibility once an upgrade is rollback-safe.

I remember discussing this. It's a really great capability.

This KEP will get us very close to being able to support it.  The gap seems to be in how we declare intent in the flags.

Say we had both a `--emulate-version=1.30` and a `--compatibile-with=1.29` flag. If we set both of them to 1.30, we'd get the behavior we want, and we'd also be able to emulate older versions as described in the KEP.

Are we OK with having two flags to control the two concepts somewhat independantly?

Are there reasonable names for two flags we could pick?  I think I could incorporate the idea into the KEP..  We're probably not complete it all in the first alpha, but it would seem reasonable to finish this all by beta.

Reply:
I think there's two distinct concepts:
1. the features you want enabled
2. the min version you want to interoperate with safely

Straw-man to represent those: `--feature-version` and `--compatibility-version`

* `--feature-version`
  * defaults to `<binaryversion>` (matching current behavior)
  * *can* set it to `<binaryversion-1>` to set feature gates and endpoint serving back to the previous version
  * could eventually set it to `<binaryversion-2>`, `<binaryversion-3>` once we retain enough historical gating
  * we could make the name of the version field in the feature gate spec tie explicitly to this:
    ```
    FeatureFoo: VersionedSpecs{
      {FeatureVersion: mustParseVersion("1.28"), ...},Alpha},
      {FeatureVersion: mustParseVersion("1.29"), ...},Beta},
      {FeatureVersion: mustParseVersion("1.30"), ...},
    },
    ```

* `--compatibility-version`
  * defaults to `<binaryversion-1>` (matching current behavior)
  * could eventually set it to `<binaryversion>` once upgrade is complete and rollback safe to accelerate enablement
  * could eventually set it to `<binaryversion-2>`, `<binaryversion-3>` once we retain enough historical gating
  * features with compatibility implications (like a new API field or relaxing validation to allow a new enum value) could include that in their feature-gate spec:
    ```
    FeatureBar: VersionedSpecs{
      {FeatureVersion: mustParseVersion("1.28"), MinCompatibilityVersion: mustParseVersion("1.28"), ...},Alpha},
      {FeatureVersion: mustParseVersion("1.29"), MinCompatibilityVersion: mustParseVersion("1.28"), ...},Beta},
      {FeatureVersion: mustParseVersion("1.30"), MinCompatibilityVersion: mustParseVersion("1.28"), ...},
    },
    ```
  * things using a gate to guard a behavior change with compatibility implications that isn't really going through the feature lifecycle could set the feature version and the min compatibility version to the same thing:
    ```
    RelaxValidationFoo: VersionedSpecs{
        {FeatureVersion: mustParseVersion("1.30"), MinCompatibilityVersion: mustParseVersion("1.30"), Default: true, ...},
    },
    ```

A given gate would only enable by default if --feature-version AND --compatibility-version were satisfied.

I don't think that is *that* different from the current state of the KEP, it just makes the implicit n-1 compatibility bit explicit, makes features think about their backcompat implications, and gives us a way to lengthen or shorten that in the future.

Thoughts?

Reply:
I prefer `--emulation-version` and `--min-compatibility-version`.

Reply:
> I prefer --emulation-version and --min-compatibility-version.

as long as we have control over the two dimensions, and we can make it as clear as possible to avoid people confusing one for the other, I don't feel that strongly about the specific names

Reply:
Works for me. I believe this new approach also reconciles the issue raised in https://github.com/kubernetes/enhancements/pull/4395#discussion_r1482193324.

Reply:
I'll put `--emulation-version` and `--min-compatibility-version` in the KEP. Further naming discussions welcome, but let's keep these names in the KEP for freeze purposes and then revisit post-freeze.

Reply:
sgtm